cornellh4i/lagos-volunteers
# Firebase auth

## Summary
1. Set up `/backend/.env` 
2. Set up `/backend/src/middleware/auth.ts` with 4 authorization functions
3. Pass authentication (only basic ones) to all existing endpoints
4. We did some experiments with creating firebase users and assigning custom claims according to the userRole ENUM, but that was aborted.

## Testing


### Procedure to test
1. Make sure you have the `.env` file. See #20 or you can DM us for the content.
2. In Postman, start a POST request with the link 
    https://www.googleapis.com/identitytoolkit/v3/relyingparty/verifyPassword?key= \<API-KEY-HERE\>
    (You need the api key for the firebase service account)
3. In POST body, write
    ```
        {
          "email":"jj@prisma.io",
          "password":"12345678",
          "returnSecureToken":true
        }
    ```
5. Send and copy the idToken
6. Make a GET request with http://localhost:8000/users/
7. Past the idToken into Authorization Bearer Token and send the request

### Result
1. Without token/or with an invalid idToken, you will have 401 unauthorized error.
2. With a valid token, you will get the result just as before. For example,
![image](https://user-images.githubusercontent.com/68068854/225164201-6bc43666-fc8f-47b7-8168-c593d4522ec1.png)
3. With a valid token but wrong CustomClaims, for example, a valid user but not an admin to make an authIfAdmin request, you will have 401 unauthorized error.

## Notes
Several things we want to mention:
1. The scope of the ticket is very broad. We believe at first that we need to implement the full-fledged testing for the different authorization, i.e. volunteer, supervisor, and admin. However, in order to test that you will need to have ADMIN or SUPERVISOR user in the first place. We realized later that we could implement assigning custom claims but the issue is when. Do we assign the roles when the user is first created? That is our first impression when we see the 10th bullet point "Modify the endpoints GET /users/:userid and POST /users/ so that they retrieve/create a user in Firebase in addition to our database." However, now we are very confused about when we should assign the roles or create the firebase users. (Update: Akin told us that we will see that part in the frontend.)
2. The endpoints are not bug-free themselves, so it would require extra work for us if we were to make sure that POST request actually work (like creating users) and the GET request is the ticket of another pair this week. We feel that it would really alleviate our pressure if we can ensure that the endpoints actually work/exist before proceeding on integrating more features. 
3. `ts-ignores` are temporarily used. The code works but ts doesn't recogize the interface extension somehow.
4. We didn't setup the Jest testsuite so it isn't working. Maybe we should remove all the "auth" before endpoints for now?
